funsor/gaussian.py

```diff
@@ -447,9 +448,92 @@ def _eager_subs_real(self, subs, remaining_subs):
         return Subs(result, remaining_subs) if remaining_subs else result
 
     def _eager_subs_affine(self, subs, remaining_subs):
-        affine = OrderedDict((k, extract_affine(v)) for k, v in subs)
-        assert affine
-        raise NotImplementedError('TODO')
+        # Extract an affine representation.
+        affine = OrderedDict()
+        for k, v in subs:
+            const, coeffs = extract_affine(v)
+            if (isinstance(const, Tensor) and
+                    all(isinstance(coeff, Tensor) for coeff, _ in coeffs.values())):
+                affine[k] = const, coeffs
+            else:
+                remaining_subs += (k, v),
+        if not affine:
+            return reflect(Subs, self, remaining_subs)
+
+        # Align integer dimensions.
+        old_int_inputs = OrderedDict((k, v) for k, v in self.inputs.items() if v.dtype != 'real')
+        tensors = [Tensor(self.info_vec, old_int_inputs),
+                   Tensor(self.precision, old_int_inputs)]
+        for const, coeffs in affine.values():
+            tensors.append(const)
+            tensors.extend(coeff for coeff, _ in coeffs.values())
+        new_int_inputs, tensors = align_tensors(*tensors, expand=True)
+        tensors = (Tensor(x, new_int_inputs) for x in tensors)
+        old_info_vec = next(tensors).data
+        old_precision = next(tensors).data
+        for old_k, (const, coeffs) in affine.items():
+            const = next(tensors)
+            for new_k, (coeff, eqn) in coeffs.items():
+                coeff = next(tensors)
+                coeffs[new_k] = coeff, eqn
+            affine[old_k] = const, coeffs
+        batch_shape = old_info_vec.data.shape[:-1]
+
+        # Align real dimensions.
+        old_real_inputs = OrderedDict((k, v) for k, v in self.inputs.items() if v.dtype == 'real')
+        new_real_inputs = old_real_inputs.copy()
+        for old_k, (const, coeffs) in affine.items():
+            del new_real_inputs[old_k]
+            for new_k, (coeff, eqn) in coeffs.items():
+                new_dim = len(eqn.split('->')[0].split(',')[1])
```

Comment:
nit: probably rename `new_dim` to something else (because it is used with a different meaning in a few lines below)

Reply:
done.